### ISSUE TITLE
Apple Silicon Unable To Build

### ISSUE THREAD

**Issue by only-cliches** [cmp-nct/ggllm.cpp#49]
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [ X] I am running the latest code. Development is very rapid so there are no tagged versions as of now.
- [ X] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [ X] I [searched using keywords relevant to my issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/filtering-and-searching-issues-and-pull-requests) to make sure that I am creating a new issue that is not already open (or closed).
- [X ] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new bug or useful enhancement to share.

# Your exact command line to replicate the issue
I tried two different commands:
```
rm -rf build; mkdir build; cd build
cmake -DLLAMA_CUBLAS=0 ..
cmake --build . --config Release
```
and
```
export LLAMA_CUBLAS=0;
make falcon_main falcon_quantize falcon_perplexity
```

# Environment and Context
```
$ make --version
GNU Make 3.81
Copyright (C) 2006  Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

This program built for i386-apple-darwin11.3.0
```

```
$ cmake --version
cmake version 3.26.3

CMake suite maintained and supported by Kitware (kitware.com/cmake).
```

```
$ /usr/bin/xcodebuild -version
Xcode 14.3.1
Build version 14E300c
```

```
$ sw_vers
ProductName:		macOS
ProductVersion:		13.4.1
BuildVersion:		22F82
```

Please provide detailed information about your computer setup. This is important in case the issue is not reproducible except for under certain specific conditions.

* Physical (or virtual) hardware you are using, e.g. for Linux:
MacBook Pro 13" with M1 Max and 64 GB of RAM.

`$ lscpu`
N/A

* Operating System, e.g. for Linux:
MacOS 13.4.1

# Steps to Reproduce
I'm unable to build this project on my Apple Silicon Mac.  Llama.cpp builds without issues.

I tried building with CMAKE and MAKE, outputs along with commands provided below:

## CMake
```
rm -rf build; mkdir build; cd build
cmake -DLLAMA_CUBLAS=0 ..
cmake --build . --config Release
```
Output:
```
[  2%] Built target BUILD_INFO
[  4%] Building C object CMakeFiles/ggml.dir/ggml.c.o
/Users/xxxxxx/Developer/ggllm.cpp/ggml.c:11037:37: error: use of undeclared identifier 'ne10'
                        ne11, ne01, ne10,
                                    ^
/Users/xxxxxx/Developer/ggllm.cpp/ggml.c:11038:37: error: use of undeclared identifier 'ne10'
                        1.0f,    y, ne10,
                                    ^
/Users/xxxxxx/Developer/ggllm.cpp/ggml.c:18287:13: warning: 'snprintf' will always overflow; destination buffer has size 128, but size argument is 512 [-Wbuiltin-memcpy-chk-size]
            snprintf(str_device_info, 512, "[CPU]");
            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/secure/_stdio.h:57:3: note: expanded from macro 'snprintf'
  __builtin___snprintf_chk (str, len, 0, __darwin_obsz(str), __VA_ARGS__)
  ^
/Users/xxxxxx/Developer/ggllm.cpp/ggml.c:19609:79: warning: format specifies type 'long long' but the argument has type 'size_t' (aka 'unsigned long') [-Wformat]
            pos += snprintf(strides + pos, sizeof(strides) - pos, "%" PRId64, tensor->nb[i]);
                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/secure/_stdio.h:57:62: note: expanded from macro 'snprintf'
  __builtin___snprintf_chk (str, len, 0, __darwin_obsz(str), __VA_ARGS__)
                                                             ^~~~~~~~~~~
/Users/xxxxxx/Developer/ggllm.cpp/ggml.c:19708:68: warning: format specifies type 'long' but the argument has type 'int64_t' (aka 'long long') [-Wformat]
            printf("Index: %d, Dimension size: %ld\n", indices[i], tensor->ne[i]);
                                               ~~~                 ^~~~~~~~~~~~~
                                               %lld
3 warnings and 2 errors generated.
make[2]: *** [CMakeFiles/ggml.dir/ggml.c.o] Error 1
make[1]: *** [CMakeFiles/ggml.dir/all] Error 2
make: *** [all] Error 2
```

## Make
```
export LLAMA_CUBLAS=0;
make falcon_main falcon_quantize falcon_perplexity
```
Output:
```
I ggllm.cpp build info:
I UNAME_S:  Darwin
I UNAME_P:  arm
I UNAME_M:  arm64
I CFLAGS:   -I.              -O3 -std=c11   -fPIC -DNDEBUG -DGGML_PERF=1 -Wall -Wextra -Wpedantic -Wcast-qual -Wdouble-promotion -Wshadow -Wstrict-prototypes -Wpointer-arith -pthread -DGGML_USE_K_QUANTS -DGGML_USE_ACCELERATE -DGGML_USE_CUBLAS -I/usr/local/cuda/include -I/opt/cuda/include -I/targets/x86_64-linux/include
I CXXFLAGS: -I. -I./examples -O3 -std=c++11 -fPIC -DNDEBUG -DGGML_PERF=1 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wno-multichar -pthread -DGGML_USE_K_QUANTS -DGGML_USE_CUBLAS -I/usr/local/cuda/include -I/opt/cuda/include -I/targets/x86_64-linux/include
I LDFLAGS:   -framework Accelerate -lcublas -lculibos -lcudart -lcublasLt -lpthread -ldl -lrt -L/usr/local/cuda/lib64 -L/opt/cuda/lib64 -L/targets/x86_64-linux/lib
I CC:       Apple clang version 14.0.3 (clang-1403.0.22.14.1)
I CXX:      Apple clang version 14.0.3 (clang-1403.0.22.14.1)

cc  -I.              -O3 -std=c11   -fPIC -DNDEBUG -DGGML_PERF=1 -Wall -Wextra -Wpedantic -Wcast-qual -Wdouble-promotion -Wshadow -Wstrict-prototypes -Wpointer-arith -pthread -DGGML_USE_K_QUANTS -DGGML_USE_ACCELERATE -DGGML_USE_CUBLAS -I/usr/local/cuda/include -I/opt/cuda/include -I/targets/x86_64-linux/include   -c ggml.c -o ggml.o
ggml.c:4262:5: error: call to undeclared function 'ggml_init_cublas'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
    ggml_init_cublas(false);
    ^
ggml.c:4319:14: error: call to undeclared function 'ggml_init_cublas'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
        if (!ggml_init_cublas(true))
             ^
ggml.c:4322:13: error: call to undeclared function 'ggml_cuda_update_gpu_status'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
            ggml_cuda_update_gpu_status(-1);
            ^
ggml.c:11037:37: error: use of undeclared identifier 'ne10'
                        ne11, ne01, ne10,
                                    ^
ggml.c:11038:37: error: use of undeclared identifier 'ne10'
                        1.0f,    y, ne10,
                                    ^
ggml.c:15751:21: error: call to undeclared function 'ggml_cuda_compute_forward'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
    bool skip_cpu = ggml_cuda_compute_forward(params, tensor);
                    ^
ggml.c:15751:21: note: did you mean 'ggml_compute_forward'?
ggml.c:15747:13: note: 'ggml_compute_forward' declared here
static void ggml_compute_forward(struct ggml_compute_params * params, struct ggml_tensor * tensor) {
            ^
ggml.c:17382:29: error: call to undeclared function 'ggml_cuda_can_mul_mat'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
                        if (ggml_cuda_can_mul_mat(node->src0, node->src1, node)) {
                            ^
ggml.c:17382:29: note: did you mean 'ggml_can_mul_mat'?
ggml.c:4132:20: note: 'ggml_can_mul_mat' declared here
static inline bool ggml_can_mul_mat(const struct ggml_tensor * t0, const struct ggml_tensor * t1) {
                   ^
ggml.c:18287:13: warning: 'snprintf' will always overflow; destination buffer has size 128, but size argument is 512 [-Wbuiltin-memcpy-chk-size]
            snprintf(str_device_info, 512, "[CPU]");
            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/secure/_stdio.h:57:3: note: expanded from macro 'snprintf'
  __builtin___snprintf_chk (str, len, 0, __darwin_obsz(str), __VA_ARGS__)
  ^
ggml.c:19609:79: warning: format specifies type 'long long' but the argument has type 'size_t' (aka 'unsigned long') [-Wformat]
            pos += snprintf(strides + pos, sizeof(strides) - pos, "%" PRId64, tensor->nb[i]);
                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/secure/_stdio.h:57:62: note: expanded from macro 'snprintf'
  __builtin___snprintf_chk (str, len, 0, __darwin_obsz(str), __VA_ARGS__)
                                                             ^~~~~~~~~~~
ggml.c:19708:68: warning: format specifies type 'long' but the argument has type 'int64_t' (aka 'long long') [-Wformat]
            printf("Index: %d, Dimension size: %ld\n", indices[i], tensor->ne[i]);
                                               ~~~                 ^~~~~~~~~~~~~
                                               %lld
3 warnings and 7 errors generated.
make: *** [ggml.o] Error 1
```

**Comment by cmp-nct** [comment]
Hi,
Well that warning about an unused ne10 I had fixed, guess it was not that unused :)
git pull again and try cmake again. That error should be fixed

The "make" problem is somewhere deeper in the makefile I suppose because the error means that GGML_USE_CUBLAS was defined despite you giving 0 as flag

**Comment by only-cliches** [comment]
Excellent, I'm able to build and run quantized Falcon models now.  Much appreciated!

**Comment by dogjamboree** [comment]
Sorry, I missed the fix here in order to get this to work on apple silicon. Could you explain it more simply?

**Comment by only-cliches** [comment]
There's a commit titled "BugFix Mac", probably includes all the changes: https://github.com/cmp-nct/ggllm.cpp/commit/4ef02f9c8e56c77e34d606b3f4437ff5a624eee8

**Comment by dogjamboree** [comment]
Does that mean it made it into the code if I pull from git? Sorry, I'm obviously pretty green when it comes to git, etc.

**Comment by only-cliches** [comment]
Yep!  If you do a `git clone` right now or use the Download link you'll have the fixed version!

You can also run `git pull` from within the repo directory if you already have this repo cloned.

These commands worked on my Apple Silicon Mac for building: (run from the root of the repo)
```
rm -rf build; mkdir build; cd build
cmake -DLLAMA_CUBLAS=0 ..
cmake --build . --config Release
```
